argon2: add parallelism

jonasmalacofilho opened the pull request:
Adds a ~~default-enabled~~ `parallel` feature, with an ~~otherwise~~ optional dependency on `rayon`, and parallelize the filling of blocks using the memory views mentioned above.

Coordinated shared access in the memory blocks is implemented with a `SegmentViewIter` iterator, which implements either `rayon::iter::ParallelIterator` or `core::iter::Iterator` and returns `SegmentView` views into the Argon2 blocks memory that are safe to be used in parallel.

The views alias in the regions that are read-only, but are disjoint in the regions where mutation happens. Effectively, they implement, with a combination of mutable borrowing and runtime checking, the cooperative contract outlined in RFC 9106. This is similar to what was suggested in #380.

To avoid aliasing mutable references into the entire buffer of blocks (which would be UB), pointers are used up to the moment where a reference (shared or mutable) into a specific block is returned. At that point, aliasing is no longer possible.

The following tests have been tried in and pass Miri (modulo unrelated warnings):

    reference_argon2i_v0x13_2_8_2
    reference_argon2id_v0x13_2_8_2

(Running these in Miri is quite slow, taking ~5 minutes each, so I only ran the most obviously relevant tests for now).

~~Finally, the alignment of `Block`s increases to 128 bytes for better prevention of false sharing on modern platforms. The new value is based on notes on [`crossbeam-utils::CachePadded`](https://github.com/crossbeam-rs/crossbeam/blob/master/crossbeam-utils/src/cache_padded.rs#L63-L79).~~

---

I also took some inspiration from an [intermediate snapshot](https://github.com/RustCrypto/password-hashes/blob/73930b2cb619643f4bb306c7a0601d423e5eecc1/argon2/src/lib.rs) of #247, before the parallel implementation was removed, as well as from an implementation without any safe abstractions I just worked on for the rust-argon2 crate (sru-systems/rust-argon2#56).

jonasmalacofilho commented:
Since the single-threaded implementation in this PR uses a simple iterator, I benchmarked the `parallel` one against the single threaded implementation in 0.5.3; that seemed like a more interesting comparison. Maybe it was what you meant anyway, otherwise let me know if you would also/still like me to benchmark both implementations in this PR.

The results show that scaling isn't great, but parallelism still makes a substantial difference. And we're lagging behind the Python (AFAIK == reference) implementation in both single threaded and parallel modes.

The benchmark results also show some single digit improvements in some 1-lane tests, but based on the other results as well as the implementation, I think that might just be (non-random) noise from things like code alignment differences between the two binaries.

<details>
<summary>Benchmark code</summary>

```rs
use std::hint::black_box;

use argon2::{
    password_hash::{rand_core::OsRng, PasswordHasher, SaltString},
    Argon2, ParamsBuilder,
};
use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion};

fn simple(c: &mut Criterion) {
    let password = b"hunter42"; // Bad password; don't actually use!
    let salt = SaltString::generate(&mut OsRng);
    let mut group = c.benchmark_group("argon2/RustCrypto/simple");
    for memory in [1, 4, 19, 46, 64, 128, 256, 512, 1024, 2048] {
        for time in [1, 2, 3] {
            for lanes in [1, 2, 4 /*, 6, 8, 12, 16*/] {
                let params = ParamsBuilder::new()
                    .p_cost(lanes)
                    .m_cost(memory * 1024)
                    .t_cost(time)
                    .build()
                    .unwrap();
                group.bench_with_input(
                    BenchmarkId::from_parameter(format!("m={memory}MiB,t={time},p={lanes},")),
                    &params,
                    |b, params| {
                        b.iter(|| {
                            Argon2::from(params)
                                .hash_password(black_box(password), black_box(&salt))
                                .unwrap()
                                .to_string()
                        })
                    },
                );
            }
        }
    }
    group.finish();
}

criterion_group!(benches, simple);
criterion_main!(benches);
```
</details>

<details>
<summary>Criterion run with argon2@0.5.3</summary>

```text
$ cargo criterion --bench argon2 -- t=1,p=1,
   Compiling benches v0.1.0 (/home/jonas/Code/rust/lab/benches)
    Finished `bench` profile [optimized] target(s) in 1.31s
argon2/RustCrypto/simple/m=1MiB,t=1,p=1,                                                                            
                        time:   [541.06 µs 541.11 µs 541.16 µs]
argon2/RustCrypto/simple/m=4MiB,t=1,p=1,                                                                             
                        time:   [2.1524 ms 2.1538 ms 2.1556 ms]
argon2/RustCrypto/simple/m=19MiB,t=1,p=1,                                                                            
                        time:   [10.867 ms 10.886 ms 10.909 ms]
argon2/RustCrypto/simple/m=46MiB,t=1,p=1,                                                                            
                        time:   [29.995 ms 30.020 ms 30.052 ms]
argon2/RustCrypto/simple/m=64MiB,t=1,p=1,                                                                            
                        time:   [41.946 ms 41.970 ms 42.000 ms]
argon2/RustCrypto/simple/m=128MiB,t=1,p=1,                                                                            
                        time:   [85.168 ms 85.213 ms 85.271 ms]
argon2/RustCrypto/simple/m=256MiB,t=1,p=1,                                                                            
                        time:   [172.12 ms 172.19 ms 172.27 ms]
argon2/RustCrypto/simple/m=512MiB,t=1,p=1,                                                                            
                        time:   [347.25 ms 347.37 ms 347.50 ms]
argon2/RustCrypto/simple/m=1024MiB,t=1,p=1,                                                                            
                        time:   [699.46 ms 699.74 ms 700.01 ms]
argon2/RustCrypto/simple/m=2048MiB,t=1,p=1,                                                                            
                        time:   [1.4036 s 1.4039 s 1.4043 s]
```
</details>

<details open>
<summary><b>Criterion run with argon2@c00251c67687 with the `parallel` feature enabled</b></summary>

```text
$ cargo criterion --bench argon2 -- t=1
   Compiling password-hash v0.6.0-rc.0
   Compiling argon2 v0.6.0-pre.1 (/home/jonas/Code/rust/password-hashes/argon2)
   Compiling benches v0.1.0 (/home/jonas/Code/rust/lab/benches)
    Finished `bench` profile [optimized] target(s) in 1.56s
argon2/RustCrypto/simple/m=1MiB,t=1,p=1,                                                                            
                        time:   [544.84 µs 545.00 µs 545.17 µs]
                        change: [+0.6580% +0.6813% +0.7072%] (p = 0.00 < 0.05)
                        Change within noise threshold.
argon2/RustCrypto/simple/m=1MiB,t=1,p=2,                                                                            
                        time:   [428.74 µs 430.90 µs 433.17 µs]
argon2/RustCrypto/simple/m=1MiB,t=1,p=4,                                                                            
                        time:   [354.73 µs 355.93 µs 357.45 µs]
argon2/RustCrypto/simple/m=4MiB,t=1,p=1,                                                                             
                        time:   [2.1444 ms 2.1452 ms 2.1462 ms]
                        change: [-0.4899% -0.4005% -0.3195%] (p = 0.00 < 0.05)
                        Change within noise threshold.
argon2/RustCrypto/simple/m=4MiB,t=1,p=2,                                                                             
                        time:   [1.4311 ms 1.4394 ms 1.4485 ms]
argon2/RustCrypto/simple/m=4MiB,t=1,p=4,                                                                             
                        time:   [969.52 µs 974.70 µs 979.90 µs]
argon2/RustCrypto/simple/m=19MiB,t=1,p=1,                                                                            
                        time:   [10.903 ms 10.913 ms 10.928 ms]
                        change: [+0.0050% +0.2430% +0.4768%] (p = 0.04 < 0.05)
                        Change within noise threshold.
argon2/RustCrypto/simple/m=19MiB,t=1,p=2,                                                                            
                        time:   [6.6484 ms 6.6716 ms 6.6959 ms]
argon2/RustCrypto/simple/m=19MiB,t=1,p=4,                                                                             
                        time:   [4.2289 ms 4.2556 ms 4.2834 ms]
argon2/RustCrypto/simple/m=46MiB,t=1,p=1,                                                                            
                        time:   [29.776 ms 29.797 ms 29.822 ms]
                        change: [-0.8689% -0.7445% -0.6258%] (p = 0.00 < 0.05)
                        Change within noise threshold.
argon2/RustCrypto/simple/m=46MiB,t=1,p=2,                                                                            
                        time:   [18.833 ms 18.879 ms 18.932 ms]
argon2/RustCrypto/simple/m=46MiB,t=1,p=4,                                                                            
                        time:   [12.582 ms 12.645 ms 12.709 ms]
argon2/RustCrypto/simple/m=64MiB,t=1,p=1,                                                                            
                        time:   [41.061 ms 41.086 ms 41.116 ms]
                        change: [-2.1982% -2.1062% -2.0118%] (p = 0.00 < 0.05)
                        Performance has improved.
argon2/RustCrypto/simple/m=64MiB,t=1,p=2,                                                                            
                        time:   [25.630 ms 25.665 ms 25.708 ms]
argon2/RustCrypto/simple/m=64MiB,t=1,p=4,                                                                            
                        time:   [16.794 ms 16.874 ms 16.957 ms]
argon2/RustCrypto/simple/m=128MiB,t=1,p=1,                                                                            
                        time:   [83.770 ms 83.822 ms 83.888 ms]
                        change: [-1.7212% -1.6326% -1.5340%] (p = 0.00 < 0.05)
                        Performance has improved.
argon2/RustCrypto/simple/m=128MiB,t=1,p=2,                                                                            
                        time:   [51.886 ms 51.977 ms 52.090 ms]
argon2/RustCrypto/simple/m=128MiB,t=1,p=4,                                                                            
                        time:   [33.603 ms 33.799 ms 33.999 ms]
argon2/RustCrypto/simple/m=256MiB,t=1,p=1,                                                                            
                        time:   [170.01 ms 170.09 ms 170.18 ms]
                        change: [-1.2878% -1.2201% -1.1534%] (p = 0.00 < 0.05)
                        Performance has improved.
argon2/RustCrypto/simple/m=256MiB,t=1,p=2,                                                                            
                        time:   [104.43 ms 104.51 ms 104.59 ms]
argon2/RustCrypto/simple/m=256MiB,t=1,p=4,                                                                            
                        time:   [67.167 ms 67.645 ms 68.127 ms]
argon2/RustCrypto/simple/m=512MiB,t=1,p=1,                                                                            
                        time:   [345.93 ms 346.04 ms 346.17 ms]
                        change: [-0.4324% -0.3836% -0.3305%] (p = 0.00 < 0.05)
                        Change within noise threshold.
argon2/RustCrypto/simple/m=512MiB,t=1,p=2,                                                                            
                        time:   [210.83 ms 211.10 ms 211.43 ms]
argon2/RustCrypto/simple/m=512MiB,t=1,p=4,                                                                            
                        time:   [134.01 ms 134.84 ms 135.69 ms]
argon2/RustCrypto/simple/m=1024MiB,t=1,p=1,                                                                            
                        time:   [691.78 ms 691.95 ms 692.13 ms]
                        change: [-1.1572% -1.1122% -1.0640%] (p = 0.00 < 0.05)
                        Performance has improved.
argon2/RustCrypto/simple/m=1024MiB,t=1,p=2,                                                                            
                        time:   [422.78 ms 423.07 ms 423.38 ms]
argon2/RustCrypto/simple/m=1024MiB,t=1,p=4,                                                                            
                        time:   [264.65 ms 265.95 ms 267.28 ms]
argon2/RustCrypto/simple/m=2048MiB,t=1,p=1,                                                                            
                        time:   [1.3907 s 1.3911 s 1.3916 s]
                        change: [-0.9526% -0.9118% -0.8749%] (p = 0.00 < 0.05)
                        Change within noise threshold.
argon2/RustCrypto/simple/m=2048MiB,t=1,p=2,                                                                            
                        time:   [848.96 ms 849.35 ms 849.74 ms]
argon2/RustCrypto/simple/m=2048MiB,t=1,p=4,
                        time:   [524.64 ms 526.04 ms 527.50 ms]                                        
```
</details>

<details>
<summary>Python `argon2` benchmarks (which AFAIK uses the reference implementation)</summary>

```text
$ python -m argon2 -t 1 -m (math "2048*1024") -p 1
Running Argon2id 100 times with:
hash_len: 32 bytes
memory_cost: 2097152 KiB
parallelism: 1 threads
time_cost: 1 iterations

Measuring...

1159.3ms per password verification

$ python -m argon2 -t 1 -m (math "2048*1024") -p 4
Running Argon2id 100 times with:
hash_len: 32 bytes
memory_cost: 2097152 KiB
parallelism: 4 threads
time_cost: 1 iterations

Measuring...

358.6ms per password verification
```
</details>

jonasmalacofilho commented on argon2/src/lib.rs:
```diff
@@ -337,126 +342,133 @@ impl<'key> Argon2<'key> {
-                    }
-
-                    let first_block = if pass == 0 && slice == 0 {
+                memory_blocks
+                    .segment_views(slice, lanes)
+                    .for_each(|mut memory_view| {
+                        let lane = memory_view.lane();
```
Please note that this `fill_blocks` diff is very noisy, due to a necessary indentation change and then rustfmt.

The only "real" changes in this function are the use of the segment view iterator (here), the accessing of memory through the segment view API instead of through indexing of the `memory_blocks` slice (bellow), and changing `memory_blocks` to be mutable (above).

jonasmalacofilho commented:
<details>
<summary>master...HEAD with parallel feature</summary>

```text
argon2i V0x10           time:   [21.324 ms 21.344 ms 21.371 ms]                          
                        change: [-0.3322% -0.1068% +0.0761%] (p = 0.34 > 0.05)
                        No change in performance detected.

argon2i V0x13           time:   [21.429 ms 21.447 ms 21.471 ms]                          
                        change: [+0.0329% +0.2197% +0.3896%] (p = 0.01 < 0.05)
                        Change within noise threshold.

argon2d V0x10           time:   [21.302 ms 21.322 ms 21.348 ms]                          
                        change: [+0.6139% +0.8010% +0.9679%] (p = 0.00 < 0.05)
                        Change within noise threshold.

argon2d V0x13           time:   [21.367 ms 21.384 ms 21.408 ms]                          
                        change: [+1.8140% +1.9978% +2.1628%] (p = 0.00 < 0.05)
                        Performance has regressed.

argon2id V0x10          time:   [21.361 ms 21.379 ms 21.405 ms]                           
                        change: [+1.2980% +1.4700% +1.6321%] (p = 0.00 < 0.05)
                        Performance has regressed.

argon2id V0x13          time:   [21.303 ms 21.320 ms 21.342 ms]                           
                        change: [+0.9147% +1.1631% +1.3556%] (p = 0.00 < 0.05)
                        Change within noise threshold.

argon2id V0x13 m=2048 t=4 p=4                                                                             
                        time:   [1.6939 ms 1.6979 ms 1.7026 ms]
                        change: [-58.795% -58.661% -58.490%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=16384 t=4 p=4                                                                            
                        time:   [11.230 ms 11.309 ms 11.391 ms]
                        change: [-67.907% -67.695% -67.447%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=65536 t=4 p=4                                                                            
                        time:   [44.778 ms 45.122 ms 45.489 ms]
                        change: [-71.067% -70.867% -70.621%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=262144 t=4 p=4                                                                            
                        time:   [172.61 ms 173.58 ms 174.61 ms]
                        change: [-72.478% -72.337% -72.127%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=32768 t=2 p=4                                                                            
                        time:   [11.964 ms 12.047 ms 12.132 ms]
                        change: [-69.521% -69.311% -69.093%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=32768 t=8 p=4                                                                            
                        time:   [45.011 ms 45.311 ms 45.623 ms]
                        change: [-69.838% -69.634% -69.434%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=32768 t=16 p=4                                                                            
                        time:   [88.879 ms 89.461 ms 90.061 ms]
                        change: [-69.861% -69.687% -69.482%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=32768 t=24 p=4                                                                            
                        time:   [133.26 ms 134.09 ms 134.93 ms]
                        change: [-69.816% -69.628% -69.446%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=2048 t=8 p=1                                                                            
                        time:   [8.1242 ms 8.1254 ms 8.1268 ms]
                        change: [+1.4099% +1.4320% +1.4529%] (p = 0.00 < 0.05)
                        Performance has regressed.

argon2id V0x13 m=2048 t=8 p=2                                                                             
                        time:   [4.8775 ms 4.9057 ms 4.9336 ms]
                        change: [-39.640% -39.331% -38.984%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=2048 t=8 p=4                                                                             
                        time:   [3.2967 ms 3.3045 ms 3.3137 ms]
                        change: [-59.213% -59.105% -58.995%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=2048 t=8 p=6                                                                             
                        time:   [2.5706 ms 2.5757 ms 2.5827 ms]
                        change: [-68.446% -68.385% -68.301%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=2048 t=8 p=8                                                                             
                        time:   [2.1205 ms 2.1339 ms 2.1500 ms]
                        change: [-73.975% -73.809% -73.631%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=2048 t=8 p=12                                                                             
                        time:   [1.8220 ms 1.8515 ms 1.8819 ms]
                        change: [-77.377% -76.954% -76.482%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=2048 t=8 p=16                                                                             
                        time:   [2.2035 ms 2.2221 ms 2.2437 ms]
                        change: [-73.287% -73.088% -72.841%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=2048 t=8 p=64                                                                             
                        time:   [2.2370 ms 2.2553 ms 2.2788 ms]
                        change: [-74.567% -74.380% -74.087%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=32768 t=4 p=1                                                                            
                        time:   [74.181 ms 74.228 ms 74.292 ms]
                        change: [-0.8519% -0.7318% -0.6115%] (p = 0.00 < 0.05)
                        Change within noise threshold.

argon2id V0x13 m=32768 t=4 p=2                                                                            
                        time:   [39.565 ms 39.759 ms 39.980 ms]
                        change: [-47.750% -47.455% -47.143%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=32768 t=4 p=4                                                                            
                        time:   [23.032 ms 23.199 ms 23.368 ms]
                        change: [-69.607% -69.389% -69.150%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=32768 t=4 p=6                                                                            
                        time:   [18.127 ms 18.171 ms 18.214 ms]
                        change: [-75.369% -75.303% -75.239%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=32768 t=4 p=8                                                                            
                        time:   [14.412 ms 14.439 ms 14.471 ms]
                        change: [-80.442% -80.403% -80.360%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=32768 t=4 p=12                                                                            
                        time:   [11.878 ms 12.021 ms 12.200 ms]
                        change: [-83.827% -83.654% -83.390%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=32768 t=4 p=16                                                                            
                        time:   [14.359 ms 14.388 ms 14.423 ms]
                        change: [-80.504% -80.462% -80.415%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=32768 t=4 p=64                                                                            
                        time:   [12.239 ms 12.285 ms 12.343 ms]
                        change: [-83.542% -83.480% -83.391%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=1048576 t=1 p=1                                                                            
                        time:   [652.11 ms 652.26 ms 652.40 ms]
                        change: [-6.4332% -6.4049% -6.3769%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=1048576 t=1 p=2                                                                            
                        time:   [337.65 ms 338.01 ms 338.40 ms]
                        change: [-51.454% -51.401% -51.345%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=1048576 t=1 p=4                                                                            
                        time:   [178.52 ms 179.41 ms 180.40 ms]
                        change: [-74.218% -74.087% -73.947%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=1048576 t=1 p=6                                                                            
                        time:   [137.57 ms 139.27 ms 141.00 ms]
                        change: [-80.074% -79.832% -79.558%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=1048576 t=1 p=8                                                                            
                        time:   [136.21 ms 136.41 ms 136.64 ms]
                        change: [-80.298% -80.265% -80.231%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=1048576 t=1 p=12                                                                            
                        time:   [119.20 ms 120.03 ms 121.02 ms]
                        change: [-82.675% -82.535% -82.391%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=1048576 t=1 p=16                                                                            
                        time:   [146.64 ms 147.06 ms 147.47 ms]
                        change: [-78.611% -78.557% -78.499%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13 m=1048576 t=1 p=64                                                                            
                        time:   [131.18 ms 131.41 ms 131.64 ms]
                        change: [-80.804% -80.771% -80.735%] (p = 0.00 < 0.05)
                        Performance has improved.
```

Note: 6-core CPU with SMT.
</details>

<details>
<summary>master...HEAD without parallel feature, default param tests only</summary>

```text
argon2i V0x10           time:   [21.365 ms 21.390 ms 21.419 ms]                          
                        change: [-0.9417% -0.7019% -0.4585%] (p = 0.00 < 0.05)
                        Change within noise threshold.

argon2i V0x13           time:   [21.523 ms 21.548 ms 21.574 ms]                          
                        change: [+0.0241% +0.2325% +0.4389%] (p = 0.03 < 0.05)
                        Change within noise threshold.

argon2d V0x10           time:   [21.201 ms 21.220 ms 21.243 ms]                          
                        change: [-0.6101% -0.4179% -0.2436%] (p = 0.00 < 0.05)
                        Change within noise threshold.

argon2d V0x13           time:   [21.403 ms 21.426 ms 21.453 ms]                          
                        change: [+0.3981% +0.6366% +0.8608%] (p = 0.00 < 0.05)
                        Change within noise threshold.

argon2id V0x10          time:   [21.241 ms 21.258 ms 21.279 ms]                           
                        change: [-1.7410% -1.5319% -1.3262%] (p = 0.00 < 0.05)
                        Performance has improved.

argon2id V0x13          time:   [21.319 ms 21.335 ms 21.355 ms]                           
                        change: [-0.9682% -0.7757% -0.5904%] (p = 0.00 < 0.05)
                        Change within noise threshold.
```
</details>

tarcieri commented:
@jonasmalacofilho if you can rebase I added `cargo careful` in #553 which should help spot issues in `unsafe` code

jonasmalacofilho commented:
@tarcieri oh, i forgot about that one. Rebased, and thanks for pointing it out!

That said, we should probably try also to add the very cheapest of tests and have it run in Miri in CI:

> That said, there is a lot of Undefined Behavior that is *not* detected by `cargo careful`; check out [Miri](https://github.com/rust-lang/miri) if you want to be more exhaustively covered. The advantage of `cargo careful` over Miri is that it works on all code, supports using arbitrary system and C FFI functions, and is much faster.

jonasmalacofilho commented:
By the way, I think there are some things I can improve in the code, but I would really appreciate a review first. And so I've kept edits to a minimum for now, so that you can actually review it.

tarcieri commented:
Yeah, Miri is tricky specifically because you can't do anything computationally expensive under it. I think we could potentially gate expensive tests under `#[cfg(not(miri))]` perhaps?

jonasmalacofilho commented:
> Yeah, Miri is tricky specifically because you can't do anything computationally expensive under it. I think we could potentially gate expensive tests under `#[cfg(not(miri))]` perhaps?

I think the 2_8_2 (t=2,m=2,p=2) tests are the cheapest in the crate, and still quite expensive... I could try adding t=1,m=8,p=2 tests and see if they execute in acceptable time in CI.

Additionally, maybe a few unit tests ensuring that allowed borrows pass in Miri, and that some known invalid borrow patterns are either impossible at compile time or caught at runtime.

BlackHoleFox commented on argon2/src/memory.rs:
```diff
@@ -0,0 +1,210 @@
+
+impl<'a> SegmentViews<'a> for &'a mut [Block] {
+    fn segment_views(&mut self, slice: usize, lanes: usize) -> SegmentViewIter<'_> {
+        // SAFETY: a pointer to the buffer or a `&mut [Block]` can never be null.
+        let blocks = unsafe { NonNull::new_unchecked(self.as_mut_ptr()) };
```
[`NonNull:from`](https://doc.rust-lang.org/stable/std/ptr/struct.NonNull.html#impl-From%3C%26mut+T%3E-for-NonNull%3CT%3E) works on mutable references, so a possible reduction in `unsafe` noise.

jonasmalacofilho replied:
Hey, thanks for the review!

One issue is that `NonNull::from(self)` would compile fine... even if someone later changed the method to just take `&self`, resulting in undefined behavior (because we would mutate through a pointer derived from a shared reference, which is UB).

That's why I opted to not rely on the `From<&mut T>` impl. But maybe I should document that in the code. Or, alternatively, I could use the impl but leave a warning that the pointer MUST be derived from a mutable reference. I think the former is harder to accidentally ignore, but I'm open to different opinions/preferences.

jonasmalacofilho replied:
We also want a `NonNull<Block>` pointer, not a `NonNull<&mut [Block]>`.

---

Regardless, there's a safe but still explicit alternative:

```suggestion
        // Note that this pointer needs to be derived from a mutable slice/reference, otherwise
        // mutating through it later would be UB.
        let blocks = NonNull::new(self.as_mut_ptr()).expect("a reference cannot dangle");
```

A quick test on Compiler Explorer shows that Rust should indeed remove the unnecessary check and panic code. So this should result in identical code as the version using `NonNull::new_unchecked`.

BlackHoleFox replied:
Nice mini investigation :D

Shame https://github.com/rust-lang/rust/issues/130823 didn't stabilize sooner but `.expect()` seems like a better middle for a security sensitive crate.

jonasmalacofilho commented:
Quick update: I ended up getting stuck trying to remove the (apparently) unrelated warnings from Miri (a warning in crossbeam and a leak due to rayon), and then I couldn't get to this PR for a few weeks.

jonasmalacofilho commented:
I removed the conflict, rebased the PR, fixed/updated the benchmarks and did some other minor cleanup.

Crossbeam-epoch doesn't currently work in Miri (see my edited comment above). Between that and the fact that even the most minimal Miri test would be super slow on GitHub free runners, I just don't think they are worth it for now. (It should be still possible to get an older toolchain and Miri and run some specific tests locally).

Is there's something else you would like me to add here?

tarcieri commented:
@jonasmalacofilho still need to go through it in detail, but there's nothing I see that's an immediate blocker

Maksych commented:
Just interesting, this or is done and waiting for merge or need some work?

tarcieri commented:
I still need to review it. Sorry it's been on the backburner since it's difficult to review due to the use of unsafe code and because we're currently working on more fundamental crates which are a blocker for another stable `argon2` release

tarcieri commented:
Apologies for not getting to this yet. I hope to get to it soon, especially as we get closer to final releases.

@jonasmalacofilho in the meantime, can you fix the conflict? Thanks

tarcieri commented on argon2/src/memory.rs:
```diff
@@ -0,0 +1,162 @@
+#![warn(
+    clippy::undocumented_unsafe_blocks,
+    clippy::missing_safety_doc,
+    unsafe_op_in_unsafe_fn
```
I don't see any `unsafe` usage in `unsafe fn`? The only `unsafe fn` I see is `SegmentView::new`

jonasmalacofilho replied:
You're right, since adding the lint I simplified the implementation significantly. Might still be worthwhile to keep the lint though.

jonasmalacofilho replied:
Actually, since the crate switched to Rust 2024 (#563), the lint is warn by default anyways. So I'll remove it.

And I missed your edit before my previous reply. In the first version of this PR there were unsafe ops in unsafe functions, and I think it's good to have a unsafe block in those cases. Rust 2024 had already made the lint warn by default, but argon2 was still on Rust 2021.

tarcieri commented on argon2/src/memory.rs:
```diff
@@ -0,0 +1,162 @@
+//! > blocks from each other. All other blocks can be referenced.
+
+#![warn(
+    clippy::undocumented_unsafe_blocks,
+    clippy::missing_safety_doc,
```
Hmm, I guess we don't have these configured at the toplevel? It would probably be good to get them there eventually

tarcieri replied:
Sure, sounds good

tarcieri reviewed (approved):
This seems plausible to me, and the benchmark improvements look impressive.

Would appreciate a second set of eyes from @newpavlov before merging.

newpavlov commented on argon2/Cargo.toml:
```diff
@@ -36,6 +37,7 @@ default = ["alloc", "password-hash", "rand"]
 alloc = ["password-hash?/alloc"]
 std = ["alloc", "password-hash?/os_rng", "base64ct/std"]
 
+parallel = ["dep:rayon", "std"]
```
IIUC we don't need `std` here since `BTreeSet` is available in `alloc`.

newpavlov reviewed (commented):
I will try to take a closer look at the implementation during this weekend.

Meanwhile, maybe we should introduce a separate feature-gated type (or generic parameter) for the parallel implementation instead of switching to it under the hood by enabling the feature?

tarcieri commented:
@newpavlov IMO it should "just work" when the feature is enabled

jonasmalacofilho commented:
I simply followed what you guys did on `balloon-hash`. Still, I think it's a simple yet good enough solution here.

Parallelism is also gated by the number of lanes, which the defender usually controls. And parallelism only serves defenders, specifically by allowing them to use higher memory and/or time parameters with acceptable latency. So if the defender doesn't want to use parallelism, they can create hashes with `p=1` (which is the currently default).
